### PR TITLE
Support for lattice type graphs

### PR DESCRIPTION
**Pull request by gkanwar**
Note: This is still waiting on merging in some frontend changes that @stephenchouca is making, but I wanted to get this pull request out for review.

This version of the branch supports stencil-assembled matrices that are both indexed and indexless. Indexless stencil-assembled matrices may only be used in gemv at the moment, as they do not combine properly with path-expression matrices.

Major changes:
-- TensorIndex supports a "Sten" kind in addition to the usual "PExpr". There are method stubs which should be implemented (and tied into index expression lowering) to allow computing values for matrix addition or multiplication.
-- Path expression Links now support a direct vertex-vertex link based on a stencil pattern. This allows generation of CSR indices based on stencils, and fits nicely into the matrix-matrix storage determination machinery.
-- Several frontend changes to support lattice sets and the map <> to <> through syntax. These will need to be updated after the frontend generics changes are merged.
-- An additional "kIndexlessStencils" global flag which chooses between indexless and indexed storage determination for stencil assembled matrices. Both cases are tested, and there is a disabled test which should soon be supported allowing combination of indexless and path-expression matrices.

### CODE REVIEW

**Comment by fredrikbk** [comment]
I've merged in the changes from @stephenchouca. Can we update this merge request?

**Comment by gkanwar** [comment]
Awesome, I'll work on merging in the new changes tonight.

**Comment by gkanwar** [comment]
Just merged in the latest generics changes. This pull request should be ready to review properly.

File: src/frontend/type_checker.cpp [hunk 1]
```diff
@@ -173,6 +173,30 @@ void TypeChecker::visit(SetType::Ptr type) {
...
+  retTypeChecked = typeCheck(type->element);
+  
+  retTypeChecked = typeCheck(type->latticePointSet) && retTypeChecked;
+  // Check lattice point set is an unstructured set
+  if (!isa<UnstructuredSetType>(type->latticePointSet->set->setDef)) {
+    std::stringstream errMsg;
+    errMsg << "expected lattice point set of Unstructured kind but got "
+           << type->latticePointSet->set->setName;
+    reportError(errMsg.str(), type->latticePointSet);
+    retTypeChecked = false;
```
Comment (stephenchouca) [inline src/frontend/type_checker.cpp]: `reportError` actually sets `retTypeChecked` to false automatically, so don't need to do it again here.

File: src/frontend/type_checker.cpp [hunk 2]
```diff
@@ -173,6 +173,30 @@ void TypeChecker::visit(SetType::Ptr type) {
...
+  }
+  // Check lattice point set is cardinality zero
+  else if (to<UnstructuredSetType>(type->latticePointSet->set->setDef)
+           ->endpoints.size() != 0) {
+    std::stringstream errMsg;
+    errMsg << "expected lattice point set of 0 cardinality, but got "
+           << to<UnstructuredSetType>(type->latticePointSet->set->setDef)
+        ->endpoints.size();
+    reportError(errMsg.str(), type->latticePointSet);
+    retTypeChecked = false;
```
Comment (stephenchouca) [inline src/frontend/type_checker.cpp]: Same as above.

File: src/frontend/type_checker.cpp [hunk 3]
```diff
@@ -269,9 +293,17 @@ void TypeChecker::visit(FuncDecl::Ptr decl) {
...
-                           to<Type>(makeTensorType(ScalarType::Type::INT)) :
-                           to<Type>(env.getSetDefinition(genericParam));
+    Type::Ptr type;
+    if (genericParam->type == GenericParam::Type::RANGE) {
+      type = to<Type>(makeTensorType(ScalarType::Type::INT));
+    }
+    else {
+      SetType::Ptr setDef = env.getSetDefinition(genericParam);
+      // We have to guard for undefined nullptr set definitions.
+      if (setDef) {
```
Comment (stephenchouca) [inline src/frontend/type_checker.cpp]: I don't think it should actually be possible for `setDef` to be null here unless there's a bug somewhere else in the type checker (presumably with `ComputeSetDefinitions`, which is responsible for setting the `setDef`'s).

File: src/frontend/type_checker.cpp [hunk 4]
```diff
@@ -1197,6 +1229,114 @@ void TypeChecker::visit(TensorReadExpr::Ptr expr) {
...
+      // TODO: No good way to check indices = #dims
+      bool typeChecked = true;
+      for (auto index : expr->indices) {
+        if (index->isSlice()) {
+          reportError("lattice point access expects integral indices", index);
+          typeChecked = false;
+        }
+      }
+
+      if (typeChecked) {
```
Comment (stephenchouca) [inline src/frontend/type_checker.cpp]: Can be replaced by `retTypeChecked`.

File: src/frontend/type_checker.cpp [hunk 5]
```diff
@@ -1551,6 +1693,36 @@ void TypeChecker::typeCheckMapOrApply(MapExpr::Ptr expr, const bool isApply) {
...
+      }
+      else if (to<LatticeLinkSetType>(throughSetType)
+               ->latticePointSet->set->setName !=
+               expr->target->setName) {
+        std::stringstream errMsg;
+        errMsg << opString << " operation can only be mapped through lattice "
+               << "link set with target " << expr->target->setName
+               << " as an endpoint";
+        reportError(errMsg.str(), expr->through);
+        retTypeChecked = false;
```
Comment (stephenchouca) [inline src/frontend/type_checker.cpp]: Unnecessary assignment.

File: src/frontend/type_checker.cpp (hunk 1 above)
Comment (gkanwar) [inline src/frontend/type_checker.cpp]: Thanks, removed here and elsewhere.

File: src/frontend/type_checker.cpp (hunk 3 above)
Comment (gkanwar) [inline src/frontend/type_checker.cpp]: You're right, this was handling an older version of how I was attempting to handle making SetType an abstract class with subclasses. I think my current iteration (with SetType::getUndefinedSetType) is neater, but let me know if there's a better design pattern.

**Comment by gkanwar** [comment]
I think I've now addressed all outstanding comments. Working now on updating the GPU backend to support this change and changes to master, but perhaps that's best done in a second pull request.

File: test/input/system/add_stencil.sim [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+    vnMat(orig,points[1,0]) = l[0,0;1,0].a;
+    vnMat(orig,points[-1,0]) = l[0,0;-1,0].a;
+end
+
+proc main
+  A = map f to springs2 reduce +;
+  B = map vonNeumann to points through springs;
+  C = A + B;
+  points.c = C*points.b;
+end
```
Comment (fredrikbk) [inline test/input/system/add_stencil.sim]: Missing newline

File: src/graph.h [hunk 7]
```diff
@@ -91,10 +91,14 @@ class ElementRef {
 // and can be passed as bound inputs to Simit programs.
 class Set {
 public:
-  Set(const std::string &name)
-      : name(name), numElements(0), endpoints(nullptr),
+  enum Kind {Unstructured, LatticeLink};
+  
+  Set(const std::string &name, Kind kind=Unstructured)
```
Comment (fredrikbk) [inline src/graph.h]: I don't think this constructor needs to take a kind(?) Either it is used to create a non-edge Set or another constructor is used. If we need this constructor it should probably be private.

Comment (gkanwar) [inline src/graph.h]: I moved this constructor to Private, and reintroduced the name-only constructor with delegation to this one. Keeping this around makes it easier to delegate from the Lattice Link constructors.

**Comment by fredrikbk** [comment]
Great work! Now Simit applies to twice as many applications

**Pull request merged**